skorch/tests/test_utils.py

```diff
@@ -28,6 +30,72 @@ def test_device_setting_cuda(self, to_tensor):
         t = to_tensor(t, device='cpu')
         assert t.device.type == 'cpu'
 
+    def tensors_equal(self, x, y):
+        """"Test that tensors in diverse containers are equal."""
+        if isinstance(x, PackedSequence):
+            return self.tensors_equal(x[0], y[0]) and self.tensors_equal(x[1], y[1])
+
+        if isinstance(x, dict):
+            return (
+                (x.keys() == y.keys()) and
+                self.tensors_equal(list(x.values()), list(y.values()))
+            )
+
+        if isinstance(x, (list, tuple)):
+            return all(self.tensors_equal(xi, yi) for xi, yi in zip(x, y))
+
+        return (x == y).all()
+
+    def parameters():
+        device = 'cpu'
+        x = torch.zeros((5, 3)).float()
+        y = torch.as_tensor([2, 2, 1])
+        z = np.arange(15).reshape(5, 3)
+        for X, expected in [
+                (x, x),
+                (y, y),
+                ([x, y], [x, y]),
+                ((x, y), (x, y)),
+                (z, torch.as_tensor(z)),
+                (
+                    {'a': x, 'b': y, 'c': z},
+                    {'a': x, 'b': y, 'c': torch.as_tensor(z)}
+                ),
+                (torch.as_tensor(55), torch.as_tensor(55)),
+                (pack_padded_sequence(x, y), pack_padded_sequence(x, y)),
+        ]:
+            yield X, expected, device
+
+        if not torch.cuda.is_available():
+            return
+
+        device = 'cuda'
+        x = x.to('cuda')
+        y = y.to('cuda')
+        for X, expected in [
+                (x, x),
+                (y, y),
+                ([x, y], [x, y]),
+                ((x, y), (x, y)),
+                (z, torch.as_tensor(z).to('cuda')),
+                (
+                    {'a': x, 'b': y, 'c': z},
+                    {'a': x, 'b': y, 'c': torch.as_tensor(z).to('cuda')}
+                ),
+                (torch.as_tensor(55), torch.as_tensor(55).to('cuda')),
+                (
+                    pack_padded_sequence(x, y),
+                    pack_padded_sequence(x, y).to('cuda')
+                ),
+        ]:
+            yield X, expected, device
```

Comment:
can't you avoid code duplication by re-using the parameters? something like

```python
devices = ['cpu']
if torch.cuda.is_available():
    devices.append('cuda')
for dev in devices:
   for p in params:
      yield (p[0], p[1].to(dev), dev)
```

Reply:
This unfortunately doesn't work because some of the parameters don't support `.to(dev)` (list, tuple, dict).

I pondered quite a bit about this parameter thing but finally settled on something that is a little bit repetitive but I hope at least easy to understand.